src/controllers/auth/complete2faSetup.ts

```diff
@@ -0,0 +1,84 @@
+import { Require_id } from 'mongoose';
+import { IUser } from '../../common/interfaces';
+import {
+	validateTimeBased2fa,
+	AppResponse,
+	getFromCache,
+	setCache,
+	toJSON,
+	decodeData,
+	generateRandom6DigitKey,
+	hashData,
+} from '../../common/utils';
+import AppError from '../../common/utils/appError';
+import { catchAsync } from '../../middlewares';
+import { UserModel } from '../../models';
+import { Request, Response } from 'express';
+import { twoFactorTypeEnum } from '../../common/constants';
+
+export const complete2faSetup = catchAsync(async (req: Request, res: Response) => {
+	const { user } = req;
+	const { token, twoFactorType } = req.body;
+
+	if (!token || !twoFactorType) {
+		throw new AppError('Token and Type is required', 400);
+	}
+
+	if (!user) {
+		throw new AppError('Unauthorized, kindly login again.');
+	}
+
+	if (user?.twoFA.active) {
+		throw new AppError('2FA is already active', 400);
+	}
+
+	if (twoFactorType === twoFactorTypeEnum.APP) {
+		const decryptedSecret = await decodeData(user.twoFA.secret as string);
```

Comment:
`user.twoFA.secret` will not be here as it is select false on the user model!

Reply:
It is coming from the cache so it'll be available